### PR TITLE
DP-1.4: Need to create forwarding groups prior to using them in a classifier

### PR DESCRIPTION
**Pull request by sgirishj**
Forwarding groups need to be created before being referenced in a classifier. DP-1.4 currently configures the classifier and adds the forwarding groups (referenced in the classifier) later. This pull request is to switch the sequence and ensure that the forwarding group is created prior to the creation of the classifier that references the groups.

### CODE REVIEW

**Comment by xw-g** [comment]
@Hobbydos Do you see any big concerns for Nokia? Thanks.

File: feature/experimental/qos/ate_tests/qos_output_queue_counters_test/qos_output_queue_counters_test.go [hunk 1]
```diff
@@ -230,6 +230,51 @@ func ConfigureQoS(t *testing.T, dut *ondatra.DUTDevice) {
...
+		targetGrpoup: "target-group-NC1",
+	}}
+
+	t.Logf("qos forwarding groups config: %v", forwardingGroups)
+	for _, tc := range forwardingGroups {
+		fwdGroup := q.GetOrCreateForwardingGroup(tc.targetGrpoup)
+		fwdGroup.SetName(tc.targetGrpoup)
+		fwdGroup.SetOutputQueue(tc.queueName)
+		queue := q.GetOrCreateQueue(tc.queueName)
+		queue.SetName(tc.queueName)
```
Comment (greg-dennis) [inline feature/experimental/qos/ate_tests/qos_output_queue_counters_test/qos_output_queue_counters_test.go:274]: This line is unnecessary -- implicit in GetOrCreateQueue

File: feature/experimental/qos/ate_tests/qos_output_queue_counters_test/qos_output_queue_counters_test.go [hunk 2]
```diff
@@ -230,6 +230,51 @@ func ConfigureQoS(t *testing.T, dut *ondatra.DUTDevice) {
 	d := &oc.Root{}
 	q := d.GetOrCreateQos()
 
+	t.Logf("Create qos forwarding groups config")
+	forwardingGroups := []struct {
```
Comment (greg-dennis) [inline feature/experimental/qos/ate_tests/qos_output_queue_counters_test/qos_output_queue_counters_test.go:234]: desc is unused and targetGroup (which is misspelled here) can just be derived from the queue name. Might as well simplify this to?

queues := []string {"BE0", "BE1", "AF1", "AF2", "AF3", "AF4", "NC1"}
t.Logf("qos forwarding groups queues: %v", queues)
for _, queue := range queues {
  targetGroup := fmt.Sprintf("target-group-%s", queue)
 ...
}

**Review by greg-dennis — CHANGES_REQUESTED** [review]

**Review by sezhang2 — CHANGES_REQUESTED** [review]
Please wait for https://github.com/openconfig/featureprofiles/pull/1100 to be merged.

Can you just create one PR for the same test code instead of multiple ones?

**Comment by sachendras** [comment]
> 

This was a request from different TLs. With small changes in different PRs, expectations are that the review process is easy and faster. Also, smaller PRs help detach changes that are unrelated and non-impacting to other implementations and can be closed sooner.

for viz: @xw-g

**Comment by sachendras** [comment]
> This was a request from different TLs. With small changes in different PRs, expectations are that the review process is easy and faster. Also, smaller PRs help detach changes that are unrelated and non-impacting to other implementations and can be closed sooner.
> 
> for viz: @xw-g

**Comment by dplore** [comment]
I agree this change is appropriate per [gnmi specification on modifying state](https://github.com/openconfig/reference/blob/master/rpc/gnmi/gnmi-specification.md#34-modifying-state).

It is ok to merge #1100 first.

**Review by sachendras — DISMISSED** [review]

**Review by sezhang2 — APPROVED** [review]

File: feature/experimental/qos/ate_tests/qos_output_queue_counters_test/qos_output_queue_counters_test.go [hunk 3]
```diff
@@ -230,111 +230,153 @@ func ConfigureQoS(t *testing.T, dut *ondatra.DUTDevice) {
...
+		desc         string
+		queueName    string
+		targetGroup  string
```
Comment (greg-dennis) [inline feature/experimental/qos/ate_tests/qos_output_queue_counters_test/qos_output_queue_counters_test.go:237]: The extraneous space between the field name and type throughout is causing the gofmt check to fail. Please fix.

File: feature/experimental/qos/ate_tests/qos_output_queue_counters_test/qos_output_queue_counters_test.go (hunk 2 above)
Comment (greg-dennis) [reply feature/experimental/qos/ate_tests/qos_output_queue_counters_test/qos_output_queue_counters_test.go:234]: When are not addressed, please say why they are not being addressed before resolving.

**Review by greg-dennis — CHANGES_REQUESTED** [review]

**Comment by LimeHat** [comment]
re: @xw-g 
> Do you see any big concerns for Nokia? Thanks.

this change seems reasonable to us

File: feature/experimental/qos/ate_tests/qos_output_queue_counters_test/qos_output_queue_counters_test.go (hunk 2 above)
Comment (sgirishj) [reply feature/experimental/qos/ate_tests/qos_output_queue_counters_test/qos_output_queue_counters_test.go:234]: Didn't address for a couple of reasons:
- Similar blocks of code are used in other places in the script. For e.g. around line # 524, for 'schedulerInts'. Better to maintain consistency across the script.
- Not directly related to the main purpose of this Pull Request (which was to fix the order of pushing different configuration blocks to resolve dependency issues)

Let me know. Thanks!

**Review by greg-dennis — APPROVED** [review]

**Review by sachendras — APPROVED** [review]

**Pull request merged**